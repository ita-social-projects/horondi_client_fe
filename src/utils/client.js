import { ApolloClient, gql, split, ApolloLink, InMemoryCache } from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { setContext } from '@apollo/client/link/context';
import fetch from 'unfetch';
import { IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
import { createUploadLink } from 'apollo-upload-client/public';
import { getMainDefinition } from '@apollo/client/utilities';

import { AUTH_ERRORS, USER_TOKENS } from '../configs';
import { getFromLocalStorage, setToLocalStorage } from '../services/local-storage.service';
import refreshAuthToken from './regenerateAuthTokenPair';

const introspectionResult = require('../fragmentTypes');

const { ACCESS_TOKEN } = USER_TOKENS;
const { ACCESS_TOKEN_IS_NOT_VALID } = AUTH_ERRORS;
const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData: introspectionResult
});

export const REACT_APP_API_URL =
  window.env && window.env.REACT_APP_API_URL
    ? window.env.REACT_APP_API_URL
    : process.env.REACT_APP_API_URL;

const wsLink = new WebSocketLink({
  uri: process.env.REACT_APP_WS_API_URL,
  options: {
    reconnect: true,
    connectionParams: {
      authToken: getFromLocalStorage(ACCESS_TOKEN)
    }
  }
});

const authLink = setContext((_, { headers }) => ({
  headers: {
    ...headers,
    token: getFromLocalStorage(ACCESS_TOKEN) || ''
  }
}));

const terminatingLink = split(
  ({ query }) => {
    const { kind, operation } = getMainDefinition(query);
    return kind === 'OperationDefinition' && operation === 'subscription';
  },
  wsLink,
  authLink.concat(createUploadLink({ uri: `${REACT_APP_API_URL}/graphql` }))
);

export const client = new ApolloClient({
  fetch,
  link: ApolloLink.from([terminatingLink]),
  cache: new InMemoryCache({
    addTypename: false,
    fragmentMatcher
  })
});

export const getItems = async (query, variables = {}) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const queryResult = await client.query({
      query: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (queryResult.data && Object.values(queryResult.data)[0]?.message) {
      const message = Object.values(queryResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await getItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return queryResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setItems = async (query, variables) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const mutationResult = await client.mutate({
      mutation: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (mutationResult.data && Object.values(mutationResult.data)[0]?.message) {
      const message = Object.values(mutationResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await setItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return mutationResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setChatMail = (query, variables) =>
  client.mutate({
    mutation: gql`
      ${query}
    `,
    variables
  });
