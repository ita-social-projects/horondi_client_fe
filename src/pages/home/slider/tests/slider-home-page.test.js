import React from 'react';
import { useQuery } from '@apollo/client';

import SliderHomePage from '../slider-home-page';

let wrapper;
const useQueryData = {
  loading: false,
  error: false,
  data: { getAllSlides: { items: [{}] } }
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: () => ({ language: 0 })
}));
jest.mock('@apollo/client');

describe('slider home page tests', () => {
  it('should match snapshot', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData
    }));

    wrapper = shallow(<SliderHomePage />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should cover other branches', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData,
      loading: true
    }));

    wrapper = shallow(<SliderHomePage />);
  });

  it('should cover rest branches', () => {
    useQuery.mockImplementation(() => ({
      ...useQueryData,
      error: {}
    }));

    wrapper = shallow(<SliderHomePage />);
  });
});
