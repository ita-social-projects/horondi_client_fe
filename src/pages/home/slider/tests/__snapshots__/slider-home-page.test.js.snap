// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slider home page tests should match snapshot 1`] = `
<div
  className="makeStyles-homeHeader-1"
  data-section-style="light"
  id="slider"
>
  <_class
    buttons={false}
    cancelOnInteraction={true}
    cssModule={null}
    fillParent={true}
    infinite={true}
    interval={5000}
    mobileTouch={true}
    onTransitionEnd={null}
    onTransitionRequest={null}
    onTransitionStart={null}
    play={true}
    rootElement="awssld"
    showTimer={true}
    timerBackgroundColor="rgba(0, 0, 0, 0.15)"
    timerHeight="6px"
  />
</div>
`;
